From 5a9710d4903e9980967b06a480bec70f10421f12 Mon Sep 17 00:00:00 2001
From: Mythri P K <mythri.p.k@intel.com>
Date: Wed, 30 Apr 2014 20:07:04 +0530
Subject: [PATCH 16/39] ASoC: soc-fw: Add pvt_data len to dapm widget

Issue: GMINL-3679

Add pvt_data_len field to dapm_widget struct, this
field is used to allocate memory for the widgets with
private data passed from user space.

Change-Id: I8642db6e11400df3bd4ead8a94b1b2364c3a7d1c
Signed-off-by: Mythri P K <mythri.p.k@intel.com>
---
 include/sound/soc-dapm.h  | 1 +
 include/uapi/sound/asoc.h | 3 +++
 sound/soc/soc-fw.c        | 3 ++-
 3 files changed, 6 insertions(+), 1 deletion(-)

diff --git a/include/sound/soc-dapm.h b/include/sound/soc-dapm.h
index b61e5d3..4f61c1e 100644
--- a/include/sound/soc-dapm.h
+++ b/include/sound/soc-dapm.h
@@ -227,6 +227,7 @@ struct snd_soc_dapm_widget {
 	struct list_head list;
 	struct snd_soc_dapm_context *dapm;
 
+	int pvt_data_len;			/* Len of widget specific data*/
 	void *priv;				/* widget specific data */
 	struct regulator *regulator;		/* attached regulator */
 	const struct snd_soc_pcm_stream *params; /* params for dai links */
diff --git a/include/uapi/sound/asoc.h b/include/uapi/sound/asoc.h
index aa777f0..56a0679 100644
--- a/include/uapi/sound/asoc.h
+++ b/include/uapi/sound/asoc.h
@@ -893,6 +893,9 @@ struct snd_soc_fw_dapm_widget {
 	__u16 event_type;
 
 	__u16 num_kcontrols;
+	__le32 reserved[12]; /*Reserved for future use */
+	__le32 pvt_data_len;
+	char pvt_data[0];
 	/*
 	 * kcontrols that relate to this widget
 	 * follow here
diff --git a/sound/soc/soc-fw.c b/sound/soc/soc-fw.c
index 6ff5f0d..6cb4ad1 100644
--- a/sound/soc/soc-fw.c
+++ b/sound/soc/soc-fw.c
@@ -1119,8 +1119,9 @@ static int soc_fw_dapm_widget_create(struct soc_fw *sfw,
 	widget.ignore_suspend = w->ignore_suspend;
 	widget.event_flags = w->event_flags;
 	widget.index = sfw->index;
+	widget.pvt_data_len = w->pvt_data_len;
 
-	sfw->pos += sizeof(struct snd_soc_fw_dapm_widget);
+	sfw->pos += (sizeof(struct snd_soc_fw_dapm_widget) + w->pvt_data_len);
 	if (w->num_kcontrols == 0) {
 		widget.num_kcontrols = 0;
 		goto widget;
-- 
1.9.1

